docs/src/pages/components/data-grid/components/components.md

```diff
@@ -7,10 +7,83 @@ components: DataGrid, XGrid
 
 <p class="description">The grid is highly customizable. Override components using the <code>components</code> prop.</p>
 
+## Installing Component
+
 As part of the customization API, the grid allows you to replace and override nested components with the `components` prop.
-The prop accepts an object of type [GridSlotsComponent](/api/data-grid/#slots) .
+The prop accepts an object of type [`GridSlotsComponent`](/api/data-grid/#slots).
+
+If you wish to pass some props in a custom component, you can do it using the `componentsProps` prop. This prop is of type `GridSlotsComponentsProps`.
+
+As an example, you could override the column menu with some props as below.
+
+```jsx
+<DataGrid
+  rows={rows}
+  columns={columns}
+  components={{
+    ColumnMenu: MyCustomColumnMenu,
+  }}
+  componentsProps={{
+    columnMenu: { background: 'red', counter: rows.length },
+  }}
+/>
+```
+
+**Note** The casing is different between the `components` and `componentsProps` prop.
+
+### Getting props
+
+While building new component slots, you might need some grid data.
+Therefore, we exposed a `useGridSlotComponentProps` hook which allows to retrieve the following props.
+
+- `state`: the current grid state.
+- `rows`: the current rows in the grid.
+- `columns`: the current columns in the grid.
+- `options`: the current set of options in the grid.
+- `apiRef`: the `GridApi` ref that allows to manipulate the grid.
+- `rootElement`: the root DOM element
+
+It can be used as below.
+
+```tsx
+function CustomRowCounter() {
+  const { rows } = useGridSlotComponentProps();
+
+  return <div> Row count: {rows.length} </div>;
+}
+```
 
-## Toolbar
+## Components
+
+The full list of overridable components can be found in the [`GridSlotsComponent`](/api/data-grid/#slots) api page.
+
+### ColumnMenu
+
+As mentioned above, the column menu is a customizable component that can be recomposed easily and customised on each column as in the demo below.
+
+{{"demo": "pages/components/data-grid/components/CustomColumnMenu.js", "bg": "inline"}}
+
+Below is our default `GridColumnMenu`.
+
+```tsx
+export const GridColumnMenu = React.forwardRef<
+  HTMLUListElement,
+  GridColumnMenuProps
+>(function GridColumnMenu(props: GridColumnMenuProps, ref) {
+  const { hideMenu, currentColumn } = props;
+
+  return (
+    <GridColumnMenuContainer ref={ref} {...props}>
+      <SortGridMenuItems onClick={hideMenu} column={currentColumn!} />
+      <GridFilterMenuItem onClick={hideMenu} column={currentColumn!} />
+      <HideGridColMenuItem onClick={hideMenu} column={currentColumn!} />
+      <GridColumnsMenuItem onClick={hideMenu} column={currentColumn!} />
+    </GridColumnMenuContainer>
+  );
+});
+```
+
+### Toolbar
```

Comment:
For the toolbar I think that showing all the imports in `CustomToolbarGrid` would be preferred. For a developer, it's a lot easier to remove one import, than to figure out which import to use for the missing component.

Reply:
Not sure I understand what that means.

Reply:
I was proposing:

```diff
diff --git a/docs/src/pages/components/data-grid/components/CustomToolbarGrid.tsx b/docs/src/pages/components/data-grid/components/CustomToolbarGrid.tsx
index 5cadc196..a13b0fe6 100644
--- a/docs/src/pages/components/data-grid/components/CustomToolbarGrid.tsx
+++ b/docs/src/pages/components/data-grid/components/CustomToolbarGrid.tsx
@@ -4,6 +4,8 @@ import {
   GridToolbarContainer,
   GridColumnsToolbarButton,
   GridFilterToolbarButton,
+  GridDensitySelector,
+  GridToolbarExport,
 } from '@material-ui/data-grid';
 import { useDemoData } from '@material-ui/x-grid-data-generator';

@@ -12,6 +14,8 @@ function CustomToolbar() {
     <GridToolbarContainer>
       <GridColumnsToolbarButton />
       <GridFilterToolbarButton />
+      <GridDensitySelector />
+      <GridToolbarExport />
     </GridToolbarContainer>
   );
 }
diff --git a/docs/src/pages/components/data-grid/components/components.md b/docs/src/pages/components/data-grid/components/components.md
index 7bbcf063..0e34c7a9 100644
--- a/docs/src/pages/components/data-grid/components/components.md
+++ b/docs/src/pages/components/data-grid/components/components.md
@@ -98,6 +98,8 @@ function CustomToolbar() {
     <GridToolbarContainer>
       <GridColumnsToolbarButton />
       <GridFilterToolbarButton />
+      <GridDensitySelector />
+      <GridToolbarExport />
     </GridToolbarContainer>
   );
 }
```